enterprise_access/utils.py

```diff
@@ -224,11 +224,16 @@ def get_normalized_metadata_for_assignment(assignment, content_metadata):
     Returns:
         dict: Normalized metadata, either for a specific course run or the advertised course run, if any.
     """
-    if not assignment.is_assigned_course_run:
-        return content_metadata.get('normalized_metadata', {})
-
+    content_key = assignment.content_key
     normalized_metadata_by_run = content_metadata.get('normalized_metadata_by_run', {})
-    return normalized_metadata_by_run.get(assignment.content_key, {})
+    # Course run-based assignments with an existing course-run based content_key
+    if assignment.is_assigned_course_run:
+        return normalized_metadata_by_run.get(content_key, {})
+    # A top level course assignment with a preferred course run key
+    if content_key := assignment.preferred_course_run_key:
+        return normalized_metadata_by_run.get(content_key, {})
+    # A top level assignment with no preferred course run.
+    return content_metadata.get('normalized_metadata', {})
```

Comment:
I believe we could potentially always rely on `preferred_course_run_key` regardless of assignment content type (course vs. course run), which would simplify this method's logic. For a run-based assignment, `preferred_course_run_key` should match the configured `content_key`; for a course-based assignment, `preferred_course_run_key` should represent the advertised run at the time of assignment allocation.

Querying the prod read-replica, there's no run-based assignments with `preferred_course_run_key=NULL` and every run-based assignment's `content_key` matches its `preferred_course_run_key`. There are ~8k course-based assignments with `preferred_course_run_key=NULL` (latest one was created May 1, which is around when this field was first introduced), which I don't think we can do much about.

What do you think of something like the following?

```python
def get_normalized_metadata_for_assignment(assignment, content_metadata):
    normalized_metadata_by_run = content_metadata.get('normalized_metadata_by_run', {})
    if preferred_course_run_key := assignment.preferred_course_run_key:
        return normalized_metadata_by_run.get(preferred_course_run_key, {})
  
    # Return current advertised course run metadata if preferred_course_run_key is NULL (i.e.,
    # impacting legacy course-based assignments created pre-May 2024).
    return content_metadata.get('normalized_metadata', {})
```

Reply:
Updated 👍🏽